pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java

```diff
@@ -18,225 +18,151 @@
-    TarArchiveInputStream debInputStream = null;
-    InputStream is = null;
-    final List<File> untaredFiles = new LinkedList<File>();
-    try {
-      is = new GzipCompressorInputStream(new BufferedInputStream(new FileInputStream(inputFile)));
-      debInputStream = (TarArchiveInputStream) new ArchiveStreamFactory().createArchiveInputStream("tar", is);
-      TarArchiveEntry entry = null;
-      while ((entry = (TarArchiveEntry) debInputStream.getNextEntry()) != null) {
-        final File outputFile = new File(outputDir, entry.getName());
-        // Check whether the untarred file will be put outside of the target output directory.
-        if (!outputFile.getCanonicalPath().startsWith(outputDirectoryPath)) {
-          throw new IOException("Tar file must not be untarred outside of the target output directory!");
+  public static List<File> untar(File inputFile, File outputDir)
+      throws IOException {
+    String outputDirCanonicalPath = outputDir.getCanonicalPath();
+    List<File> untarredFiles = new ArrayList<>();
+    try (InputStream fileIn = Files.newInputStream(inputFile.toPath());
+        InputStream bufferedIn = new BufferedInputStream(fileIn);
+        InputStream gzipIn = new GzipCompressorInputStream(bufferedIn);
+        ArchiveInputStream tarGzIn = new TarArchiveInputStream(gzipIn)) {
+      ArchiveEntry entry;
+      while ((entry = tarGzIn.getNextEntry()) != null) {
+        String entryName = entry.getName();
+        String[] parts = StringUtils.split(entryName, ENTRY_NAME_SEPARATOR);
+        File outputFile = outputDir;
+        for (String part : parts) {
+          outputFile = new File(outputFile, part);
         }
         if (entry.isDirectory()) {
-          // creating directory
-          LOGGER.debug(String.format("Attempting to write output directory %s.", outputFile.getAbsolutePath()));
-          if (!outputFile.exists()) {
-            // does not exist
-            LOGGER.debug(String.format("Attempting to create output directory %s.", outputFile.getAbsolutePath()));
-            try {
-              // create the directory including any non-existent parent directories
-              Files.createDirectories(outputFile.toPath());
-            } catch (Exception e) {
-              LOGGER.error("Caught exception while creating directory: {}, error: {}", outputFile.getAbsolutePath(), e);
-              throw e;
-            }
-          } else {
-            // directory already exists
-            LOGGER.error("The directory already there. Deleting - " + outputFile.getAbsolutePath());
-            FileUtils.deleteDirectory(outputFile);
+          if (!outputFile.getCanonicalPath().startsWith(outputDirCanonicalPath)) {
+            throw new IOException(String
+                .format("Trying to create directory: %s outside of the output directory: %s", outputFile, outputDir));
+          }
+          if (!outputFile.isDirectory() && !outputFile.mkdirs()) {
```

Comment:
Since entry is a directory and each part leading up to the entry is also a directory, outputFile is also a directory right since `outputFile = new File(outputFile, part)`;

Reply:
`outputFile` is the file/dir pointed by the entry, so if entry is directory, then `outputFile` is directory